[nn] add remove_duplicate flag to named_parameters (#759)

samdow opened the pull request:
Summary:
X-link: https://github.com/pytorch/torchrec/pull/759

Since the remove_duplicate flag was added to named_buffers in D39493161 (https://github.com/pytorch/pytorch/commit/c12f829cce29eb6971094a9bbb0f8971aed86f5c), this adds the same flag to named_parameters

Test Plan:
python test/test_nn.py -k test_buffers_and_named_buffers

OSS Tests

Differential Revision: D40801899



cc @ezyang @gchanan

albanD commented on test/test_nn.py:
```diff
@@ -805,6 +805,18 @@ def __init__(self):
         self.assertEqual(names(m.named_buffers(remove_duplicate=False)),
                          ["buffer1", "buffer2"])
 
+        class M(nn.Module):
```
Could you also have a test for weight being shared because the same Module is re-used
```
self.mod1 = nn.Linear()
self.mod2 = self.mod1
```

samdow replied:
Added

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1958,7 +1958,12 @@ def parameters(self, recurse: bool = True) -> Iterator[Parameter]:
+            self,
+            prefix: str = '',
+            recurse: bool = True,
+            remove_duplicate: bool = True
```
I guess we can't make these kwarg only? :(

albanD reviewed (approved):
Sounds good. Just an extra test.

samdow committed:
[nn] add remove_duplicate flag to named_parameters (#88090)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88090

X-link: https://github.com/pytorch/torchrec/pull/759

Since the remove_duplicate flag was added to named_buffers in D39493161 (https://github.com/pytorch/pytorch/commit/c12f829cce29eb6971094a9bbb0f8971aed86f5c), this adds the same flag to named_parameters

Test Plan:
python test/test_nn.py -k test_buffers_and_named_buffers

OSS Tests

Reviewed By: albanD

Differential Revision: D40801899

fbshipit-source-id: 163bfd5052c42b9bd87cb35931d9f7be627e3256
